man1/nodelint.1
.\" Generated with Ronnjs/v0.1
.\" http://github.com/kapouer/ronnjs/
.
.TH "NODELINT" "1" "November 2011" "" ""
.
.SH "NAME"
\fBnodelint\fR \-\- Run JSLint from the command\-line under node\.js
.
.SH "SYNOPSIS"
.
.nf
nodelint [options] <file\-or\-directory> [<file\-or\-directory> \.\.\.]
.
.fi
.
.SH "DESCRIPTION"
The nodelint command\-line tool allows you to check for problems and ensure
.
.br
the code quality of your JavaScript files using JSLint\.
.
.P
It is completely extensible so you can use your own custom JSLint config or
.
.br
even use custom reporters that better integrate with your quality assurance
.
.br
framework\.
.
.SH "OPTIONS"
  \fB\-\-reporter FILE\fR:
.
.br
      Override the default reporter with your own custom module\. See
.
.br
      the \fIexamples/reporters\fR directory for custom reporters that come
.
.br
      bundled with nodelint\.
.
.P
  \fB\-\-config FILE\fR:
.
.br
      Override the default \fIconfig\.js\fR with your own config file\.
.
.P
  \fB\-h\fR, \fB\-\-help\fR:
.
.br
      Display the help and exit\.
.
.P
  \fB\-v\fR, \fB\-\-version\fR:
.
.br
      Output version information and exit\.
.
.P
  \fB<file\-or\-directory>\fR:
      You can run nodelint on specific files or on all \fI*\.js\fR files inside
.
.br
      a directory\.
.
.SH "CONFIG"
You can customise the JSLint options by modifying the default config\.js
.
.br
file or by providing your own config file with the \fI\-\-config\fR parameter:
.
.IP "" 4
.
.nf
nodelint \-\-config path/to/custom\.js file1\.js file2\.js \.\.\.
.
.fi
.
.IP "" 0
.
.P
For example, if the default config\.js looks like:
.
.IP "" 4
.
.nf
var options = {
    adsafe       : false,
    bitwise      : true,
    error_prefix : "\\u001b[1m",
    error_suffix : ":\\u001b[0m "
};
.
.fi
.
.IP "" 0
.
.P
And your custom\.js looks like:
.
.IP "" 4
.
.nf
var options = {
    bitwise      : false,
    browser      : false
};
.
.fi
.
.IP "" 0
.
.P
Then the final options will be:
.
.IP "" 4
.
.nf
var options = {
    adsafe       : false,
    bitwise      : false,
    browser      : false,
    error_prefix : "\\u001b[1m",
    error_suffix : ":\\u001b[0m "
};
.
.fi
.
.IP "" 0
.
.SH "JSLINT OPTIONS"
.
.IP "\(bu" 4
adsafe:
.
.br
True if ADsafe  rules should be enforced\. See http://www\.ADsafe\.org/\.
.
.IP "\(bu" 4
bitwise:
.
.br
True if bitwise operators should not be allowed\.
.
.IP "\(bu" 4
browser:
.
.br
True if the standard browser globals should be predefined\.
.
.IP "\(bu" 4
cap:
.
.br
True if upper case HTML should be allowed\.
.
.IP "\(bu" 4
css:
.
.br
True if CSS workarounds should be tolerated\.
.
.IP "\(bu" 4
debug:
.
.br
True if debugger statements should be allowed\.
.
.br
Set this option to false before going into production\.
.
.IP "\(bu" 4
devel:
.
.br
True if browser globals that are useful in development
.
.br
(console, alert, \.\.\.) should be predefined\.
.
.IP "\(bu" 4
eqeqeq:
.
.br
True if === should be required\.
.
.IP "\(bu" 4
es5:
.
.br
True if ES5 syntax should be allowed\.
.
.IP "\(bu" 4
evil:
.
.br
True if eval should be allowed\.
.
.IP "\(bu" 4
forin:
.
.br
True if unfiltered for in statements should be allowed\.
.
.IP "\(bu" 4
fragment:
.
.br
True if HTML fragments should be allowed\.
.
.IP "\(bu" 4
immed:
.
.br
True if immediate function invocations must be wrapped in parens
.
.IP "\(bu" 4
indent:
.
.br
The number of spaces used for indentation (default is 4)
.
.IP "\(bu" 4
laxbreak:
.
.br
True if statement breaks should not be checked\.
.
.IP "\(bu" 4
maxerr:
.
.br
The maximum number of warnings reported (default is 50)
.
.IP "\(bu" 4
maxlen:
.
.br
The maximum number of characters in a line
.
.IP "\(bu" 4
nomen:
.
.br
True if names should be checked for initial or trailing underbars
.
.IP "\(bu" 4
newcap:
.
.br
True if Initial Caps must be used with constructor functions\.
.
.IP "\(bu" 4
on:
.
.br
True if HTML event handlers should be allowed\.
.
.IP "\(bu" 4
onevar:
.
.br
True if only one var statement per function should be allowed\.
.
.IP "\(bu" 4
passfail:
.
.br
True if the scan should stop on first error\.
.
.IP "\(bu" 4
plusplus:
.
.br
True if ++ and \-\- should not be allowed\.
.
.IP "\(bu" 4
predef:
.
.br
An array of strings (comma separated), the names of predefined global variables\.
.
.br
predef is used with the option object, but not with the /\fIjslint \fR/ comment\.
.
.br
Use the var statement to declare global variables in a script file\.
.
.IP "\(bu" 4
regexp:
.
.br
True if \. and [^\.\.\.] should not be allowed in RegExp literals\.
.
.br
These forms should not be used when validating in secure applications\.
.
.IP "\(bu" 4
rhino:
.
.br
True if the Rhino environment globals should be predefined\.
.
.IP "\(bu" 4
safe:
.
.br
True if the safe subset rules are enforced\. These rules are used by ADsafe\.
.
.br
It enforces the safe subset rules but not the widget structure rules\.
.
.IP "\(bu" 4
strict:
.
.br
True if the ES5 "use strict"; pragma is required\. Do not use this option carelessly\.
.
.IP "\(bu" 4
sub:
.
.br
True if subscript notation may be used for expressions better expressed in dot notation\.
.
.IP "\(bu" 4
undef:
.
.br
True if variables must be declared before used\.
.
.IP "\(bu" 4
white:
.
.br
True if strict whitespace rules apply\.
.
.IP "\(bu" 4
widget:
.
.br
True if the Yahoo Widgets globals should be predefined\.
.
.IP "\(bu" 4
windows:
.
.br
True if the Windows globals should be predefined\.
.
.IP "" 0
.
.SH "AUTHORS"
Written by Tav and other nodelint contributors\.
.
.br
Contributors list: \fIhttps://github\.com/tav/nodelint/contributors\fR\|\.
.
.SH "REPORTING BUGS"
Report nodelint bugs to \fIhttps://github\.com/tav/nodelint/issues\fR\|\.
.
.SH "COPYRIGHT"
Nodelint has been released into the Public Domain by its Authors\.
.
.SH "SEE ALSO"
node(1)


nodelint.js
#! /usr/bin/env node

/*
 * This is a command line runner of jslint for Node.js
 *
 * Changes released into the Public Domain by tav <tav@espians.com>
 * Support for config/multiple files added by Corey Hart <corey@codenothing.com>
 *
 * Adapted from rhino.js, Copyright (c) 2002 Douglas Crockford
 * Adapted from posixpath.py in the Python Standard Library.
 *
 */
 
/*global JSLINT, options */
/*jslint evil: true, regexp: false, indent: 2 */

(function () {
  var fs = require('fs'),
      path = require('path'),
      sys = require('sys'),
      SCRIPT_DIRECTORY,
      DEFAULT_CONFIG_FILE,
      params,
      files,
      config_file,
      config_param_found,
      reporter_file,
      reporter_param_found;

// -----------------------------------------------------------------------------
// default reporter for printing to a console
// -----------------------------------------------------------------------------

  function reporter(results, error_prefix, error_suffix) {
    var error_regexp = /^\s*(\S*(\s+\S+)*)\s*$/,
        i,
        len = results.length,
        str = '',
        error;
        
    for (i = 0; i < len; i += 1) {
      error = results[i].error;
      str += error_prefix + results[i].file  + ', line ' + error.line +
             ', character ' + error.character + error_suffix +
              error.reason + '\n' +
              (error.evidence || '').replace(error_regexp, "$1") + '\n';
    }
    str += len + ' error' + ((len === 1) ? '' : 's');
    sys.error(str);
  }

// -----------------------------------------------------------------------------
// load jslint itself and set the path to the default config file
// -----------------------------------------------------------------------------
  
  SCRIPT_DIRECTORY = path.dirname(__filename);
  DEFAULT_CONFIG_FILE = path.join(SCRIPT_DIRECTORY, 'config.js');

  eval(fs.readFileSync(path.join(SCRIPT_DIRECTORY, 'jslint/jslint.js'), 'utf8'));
  
// -----------------------------------------------------------------------------
// skript main funktion
// -----------------------------------------------------------------------------

  function lint(files, default_config_file, config_file, reporter_file) {
    var retval = 0,
        results = [],
        error_prefix,
        error_suffix,
        option_name,
        real_options;
        
    if (!files.length) {
      sys.puts(
          "Usage: nodelint.js file.js [file2 file3 ...] [options]\n" +
          "Options:\n\n" +
          "  --config FILE  the path to a config.js file with JSLINT options\n" +
          "  --reporter FILE  optional path to a reporter file to customize the output"
        );
      return 1;
    }
    
    eval(fs.readFileSync(default_config_file, 'utf8'));
    
    if (typeof options === 'undefined') {
      sys.puts("Error: there's no `options` variable in the default config file.");
      return 1;
    }
    
    real_options = options;
    
    if (typeof config_file !== 'undefined') {
      eval(fs.readFileSync(config_file, 'utf8'));

      if (typeof options === 'undefined') {
        sys.puts("Error: there's no `options` variable in the config file.");
        return 1;
      }
      
      for (option_name in options) {
        if (typeof option_name === 'string') {
          real_options[option_name] = options[option_name];
        }
      }
    }
    
    if (typeof reporter_file !== 'undefined') {
      eval(fs.readFileSync(reporter_file, 'utf8'));
    }
    
    error_prefix = real_options.error_prefix;
    error_suffix = real_options.error_suffix;
    
    files.forEach(function (file) {
      var source,
          i,
          error;

      try {
        source = fs.readFileSync(file, 'utf8');
      } catch (err) {
        sys.puts("Error: Opening file <" + file + ">");
        sys.puts(err + '\n');
        retval = 1;
        return;
      }

      // remove any shebangs
      source = source.replace(/^\#\!.*/, '');

      if (!JSLINT(source, real_options)) {
        for (i = 0; i < JSLINT.errors.length; i += 1) {
          error = JSLINT.errors[i];
          if (error) {
            results.push({file: file, error: error});
          }
        }
        retval = 2;
        return;
      }
    });
    
    reporter(results, error_prefix, error_suffix);
    return retval; 
  }
  
// -----------------------------------------------------------------------------
// run the file as a script if called directly, i.e. not imported via require()
// -----------------------------------------------------------------------------

  params = process.ARGV.splice(2);
  files = [];
   
  // a very basic pseudo --options parser
  params.forEach(function (param) {
    if (param.slice(0, 9) === "--config=") {
      config_file = param.slice(9);
    } else if (param === '--config') {
      config_param_found = true;
    } else if (config_param_found) {
      config_file = param;
      config_param_found = false;
    } else if (param === '--reporter') {
      reporter_param_found = true;
    } else if (reporter_param_found) {
      reporter_file = param;
      reporter_param_found = false;
    } else if ((param === '--help') || (param === '-h')) {
    } else {
      files.push(param);
    }
  });
  
  process.exit(lint(files, DEFAULT_CONFIG_FILE, config_file, reporter_file));
  
}());
